server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GC.scala

```diff
@@ -0,0 +1,187 @@
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.server.blob.deduplication
+
+import org.apache.james.blob.api.BlobId
+
+/**
+ * Isolating and grouping Events
+ */
+sealed abstract class Generation extends Comparable[Generation] {
+  def previous: Generation
+  def previous(times: Long): Generation =
+    (0L until times).foldLeft(this)((generation, _) => generation.previous)
+
+  def next: Generation
+  def next(times: Long): Generation =
+    (0L until times).foldLeft(this)((generation, _) => generation.next)
+
+  /**
+   * List all generations the GC is able to collect
+   */
+  def collectibles(targetedGeneration: Generation): Set[Generation] =
+    Generation.range(this, targetedGeneration.previous(GC.temporization)).toSet
+
+  def <(that: Generation): Boolean = compareTo(that) < 0
+  def <=(that: Generation): Boolean = compareTo(that) <= 0
+  def >(that: Generation): Boolean = compareTo(that) > 0
+  def >=(that: Generation): Boolean = compareTo(that) >= 0
+}
+
+object Generation {
+  val first: Generation = apply(0)
+
+  def apply(id: Long): Generation = {
+    if (id < 0) {
+      NonExistingGeneration
+    } else {
+      ValidGeneration(id)
+    }
+  }
+
+  def range(start: Generation, end: Generation): Seq[Generation] = (start, end) match {
+    case (NonExistingGeneration, NonExistingGeneration) => Seq(NonExistingGeneration)
+    case (ValidGeneration(_), NonExistingGeneration) => Nil
+    case (NonExistingGeneration, ValidGeneration(id)) =>  NonExistingGeneration +: (0L to id).map(Generation.apply)
+    case (ValidGeneration(id1), ValidGeneration(id2)) => (id1 to id2).map(Generation.apply)
+  }
+}
+
+/**
+ * Generation which has existed
+ */
+case class ValidGeneration(id: Long) extends Generation {
+  override def previous: Generation = Generation(id - 1)
+
+  override def next: Generation = copy(id + 1)
+
+  override def compareTo(t: Generation): Int = t match {
+    case NonExistingGeneration => 1
+    case that: ValidGeneration => id.compareTo(that.id)
+  }
+
+}
+
+/**
+ * NullObject for the initialisation of the GC
+ */
+case object NonExistingGeneration extends Generation {
+  override def previous: Generation = NonExistingGeneration
+
+  override def next: Generation = Generation.first
+
+  override def compareTo(t: Generation): Int = t match {
+    case NonExistingGeneration => 0
+    case _: ValidGeneration => -1
+  }
+}
+
+/**
+ * A run of the GC regarding a Set of Generations
+ */
+case class Iteration(id: Long, processedGenerations: Set[Generation], lastGeneration: Generation) {
+  def next(generations: Set[Generation], lastGeneration: Generation): Iteration = Iteration(id + 1, generations, lastGeneration)
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(), NonExistingGeneration)
+}
+
+case class ExternalID(id: String)
```

Comment:
What does this case class represents?

Reply:
any non GC-managed id able to reference a BlobId

Reply:
Ok like a messageId, attachmentId etc...